subsys/dfu/img_util/flash_block.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __FLASH_IMG_H__
+#define __FLASH_IMG_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct flash_img_contex {
+	u8_t buf[CONFIG_IMG_BLOCK_BUF_SIZE];
+	struct device *dev;
+	u32_t bytes_written;
+	u16_t buf_bytes;
+};
+
+/**
+ * @brief Initialize of context needed for proceed the image writing.
+ *
+ * @param cb context to be initialized
+ * @param dev flash driver to be used to proceed the writing
+ */
+void img_writer_init(struct flash_img_contex *cb, struct device *dev);
+
+/**
+ * @brief Extract count of bytes written to the firmware image.
+ *
+ * @param cb context
+ *
+ * @retval Count of bytes written to the image flash.
+ */
+u32_t img_written_size_get(struct flash_img_contex *cb);
+
+/**
+ * @brief  Process input buffers to be written to the 1. image bank flash
+ * memory in single blocks. Will store remainder between calls.
+ *
+ * A final call to this function with finished set to true
+ * will write out the remaining block buffer to flash.
+ *
+ * @param cb context
+ * @param data data to write
+ * @param len Number of bytes to write
+ * @param finished when true this forces any buffered
```

Comment:
Since this is about flushing a buffer, maybe rename this argument "flush" instead for submission upstream?

I also think that the documentation here should add some extra discussion about what it means to flush these bytes. 

The routine will pad the buffer out with 0xff up to the next CONFIG_IMG_BLOCK_BUF_SIZE boundary before writing, and not all flash hardware allows writing 0xff more than once. So the entire flash device up to the next multiple of CONFIG_IMG_BLOCK_BUF_SIZE is no longer safe to write until the next erase. This may be more bytes than what the user expects, so it seems like a good idea to make a note of that in the documentation.